src/typechecker.erl

```diff
@@ -608,25 +608,36 @@ glb_ty(_Ty1, _Ty2, _A, _TEnv) -> {type(none), constraints:empty()}.
 %% * Replace built-in type synonyms
 %% * Flatten unions and merge overlapping types (e.g. ranges) in unions
 -spec normalize(type(), TEnv :: #tenv{}) -> type().
-normalize({type, _, record, [{atom, _, Name}|Fields]}, TEnv) when length(Fields) > 0 ->
-    NormFields = [type_field_type(FieldName, normalize(Type, TEnv)) 
+normalize(Ty, TEnv) ->
+    normalize_rec(Ty, TEnv, sets:new()).
+
+%% The third argument is a set of user types that we've already unfolded.
+%% It's important that we don't keep unfolding such types because it will
+%% lead to infinite recursion.
+normalize_rec({type, _, record, [{atom, _, Name}|Fields]}, TEnv, Unfolded)
+  when length(Fields) > 0 ->
```

Comment:
Don't we need a base case to actually stop recursion? (not as a guard expression but you know what I mean)

```Erlang
normalize_rec(Type, _TEnv, Unfolded) when sets:is_element(Type, Unfolded) ->
    Type;
```

Reply:
Ah, sorry, this is about recursive user-defined types only.